pubtools/_pulp/tasks/push/items/base.py

```diff
@@ -281,11 +300,42 @@ def refresh_after_copy(_):
 
             return f
 
+        # A helper function to re-try copying of failed items
+        #  Since the association phase can take a while and the items can be
+        #  moved to other repos meanwhile, we fetch up-to-date info about the
+        #  items from pulp and re-try the failing cases
+        def retry_failed_items(_):
+            copy_results = []
+            for i in range(association_retry_count):
```

Comment:
I don't think the loop is going to work as desired in the case where more than 1 retry happens.  Correct me if I'm wrong but I think it'll work like this:

- first iter (i=0)
- get state of all pulp items
- let's say that enqueue_items returns non-empty copy_results, so some copies are being retried
- we log "Retrying association"
- then *we do not wait for copy to complete*, so we immediately continue on to...
- second iter (i=1)
- get state of all pulp items
- given that we've only just started a new copy and didn't wait for it to complete, the state is probably going to be the same as it was at i=0, so we're going to retry the same copy again.

By my reading this loop is always going to queue retries up to the configured retry count as fast as possible, or have I missed something?

If the loop is aiming to achieve "retry copy until it succeeds" I'd expect somewhere in the loop body it'll have to wait on the results, probably just after the log message?  Something like `f_sequence(copy_results).result()` ?

Note it'd generally be a bad practice for a function like this to block and call `.result()`, but I think it's OK in this case since retrying should be a niche scenario.

Reply:
Thank you for the comment! I have completely missed that. 

I have added `f_sequence(copy_results).result()` to the loop.